doc/syslog/syslog-rate-limit-design.md

```diff
@@ -0,0 +1,376 @@
+ |:---:|:-----------:|:------------------:|-----------------------------------|
+ | 0.1 |             |      Junchao Chen  | Initial version                   |
+
+### Scope
+
+This document is the design document for syslog message rate limit configuration per container.
+
+### Definitions/Abbreviations
+
+N/A
+
+### Overview
+
+Logging in SONiC is organized with rsyslogd. Each container has its own rsyslogd instance plus a daemon running on host side. The rsyslogd instance which is running on the host is used to collect the messages from within containers and store them at certain path (E.g. /var/log/syslog). Rsyslog config file are generated from templates:
+
+- Container scope:
+  - Multi ASIC: https://github.com/Azure/sonic-buildimage/blob/master/files/image_config/rsyslog/rsyslog-container.conf.j2
+  - Single ASIC: https://github.com/Azure/sonic-buildimage/blob/master/dockers/docker-base/etc/rsyslog.conf
+- Host scope: https://github.com/Azure/sonic-buildimage/blob/master/files/image_config/rsyslog/rsyslog.conf.j2
+
+Currently, each container has hardcoded message rate limiting to avoid receiving flooded log messages:
+
+```
+$SystemLogRateLimitInterval 300
+$SystemLogRateLimitBurst 20000
+```
+There is no rate limiting configured on host side for now.
+
+The SystemLogRateLimitInterval determines the amount of time that is being measured for rate limiting. The SystemLogRateLimitBurst defines the amount of messages, that have to occur in the time limit of SystemLogRateLimitInterval, to trigger rate limiting. For example, SystemLogRateLimitInterval=300, SystemLogRateLimitBurst=20000, it means that if one daemon generate more than 20000 messages in 300 seconds, rsyslogd will start to drop messages after that(FIFO).
+
+This feature allows user to configure SystemLogRateLimitInterval and SystemLogRateLimitBurst for host, containers.
+
+### Requirements
+
+- Support syslog message rate limit configuration for host and containers
+- New CLI for rate limit configuration
+- Have rate limit configuration persistent
+- Default rate limit shall be applied if no configuration provided
+- Both multi ASIC and single ASIC platform shall use rsyslog-container.conf.j2 as template
+- APP extension shall be able to declare its capability of this feature
+- APP extension is responsible for reading/listening rate limit configuration from CONFIG DB and set rsyslog configuration accordingly.
+
+### Architecture Design
+
+rsyslog.conf.j2 and rsyslog-container.conf.j2 shall be extended to accept template variable for SystemLogRateLimitInterval and SystemLogRateLimitBurst. rsyslog.conf for single ASIC platform will be removed and replaced by rsyslog-container.conf.j2.
+
+Host side:
+
+![host_config_on_fly](/doc/syslog/images/host_rate_limit_config_flow.svg).
+
+Change syslog message rate limit while container is running:
+
+![container_config_on_fly](/doc/syslog/images/container_rate_limit_config_flow.svg).
```

Comment:
@Junchao-Mellanox the container flow diagram shows only the modification flow. Who will put container initial rsyslogd configuration in that case?

Reply:
There is no change for container startup flow.

Reply:
There is no change for container initial flow. Existing container init flow will render rsyslog-container.conf.j2 and "docker cp" it to container. So, I did not draw the chart here.

Reply:
Chart provided